test/tests/functional/pbs_ralter.py

```diff
@@ -1516,3 +1550,254 @@ def test_failed_ralter(self):
+        shift = 10
+        rid, start, end = self.submit_and_confirm_reservation(offset, duration)
+
+        self.alter_a_reservation(rid, start, end, shift=shift,
+                                 alter_s=True, a_duration=30, check_log=False)
+
+        t_duration, t_start, t_end = self.get_resv_time_info(rid)
+
+        self.assertEqual(t_end, end+20)
+        self.assertEqual(t_start, start+10)
+        self.assertEqual(t_duration, 30)
+
+        # Submit a job to the reservation and change its start time.
+        self.submit_job_to_resv(rid)
+        temp_end = t_end
+        temp_start = t_start
+
+        self.alter_a_reservation(rid, temp_start, temp_end, shift=shift,
+                                 alter_s=True, sequence=2,
+                                 a_duration=40, check_log=False)
+        t_duration, t_start, t_end = self.get_resv_time_info(rid)
+        self.assertEqual(t_end, temp_end+20)
+        self.assertEqual(t_start, temp_start+10)
+        self.assertEqual(t_duration, 40)
+
+    def test_adv_res_dur_after_start(self):
+        """
+        Test that duration can be changed after the reservation starts.
+        Test that if the duration changes endtime of the reservation to an
+        already passed time, the reservation is deleted
+        """
+        offset = 10
+        duration = 20
+        shift = 10
+        rid, start, end = self.submit_and_confirm_reservation(offset, duration)
+
+        self.check_resv_running(rid, offset)
+
+        self.alter_a_reservation(rid, start, end, a_duration=30,
+                                 confirm=False, check_log=False)
+
+        t_duration, t_start, t_end = self.get_resv_time_info(rid)
+        self.assertEqual(t_duration, 30)
+
+        time.sleep(5)
+        attr = {'reserve_duration': 5}
+        self.server.alterresv(rid, attr)
+        self.server.log_match(rid + ";Reservation alter denied",
+                              id=rid, interval=2)
+        rid = rid.split('.')[0]
+        self.server.log_match(rid + ";deleted at request of pbs_server",
+                              id=rid, interval=2)
+
+    def test_adv_resv_endtime_starttime_dur_together(self):
+        """
+        Test that all three end, start and duration can be changed together
+        """
+        offset = 20
+        duration = 20
+        shift = 10
+        rid, start, end = self.submit_and_confirm_reservation(offset, duration)
+        new_start = self.bu.convert_seconds_to_datetime(start + shift)
+        new_end = self.bu.convert_seconds_to_datetime(end + shift)
+        new_duration = self.bu.convert_seconds_to_datetime(duration + 10)
+
+        try:
+            attr = {'reserve_start': new_start, 'reserve_end': new_end,
+                    'reserve_duration': new_duration}
+            self.server.alterresv(rid, attr)
+        except PbsResvAlterError as e:
+            self.assertTrue("pbs_ralter: Bad time specification(s)" in
+                            e.msg[0])
```

Comment:
@vstumpf has a better way of doing this.  It's with assertRaises().  Could you provide suggested code @vstumpf ?

Reply:
Sure, it could be:
```Python
with self.assertRaises(PbsResvAlterError) as e:
    attr = {'reserve_start': new_start, 'reserve_end': new_end,
        'reserve_duration': new_duration}
    self.server.alterresv(rid, attr)
self.assertIn('pbs_ralter: Bad time specification(s)', e.msg[0])

Reply:
You should also make sure that the reservation has the original start/end/duration

Reply:
I'd move the setting of attr outside of the with, but it really doesn't matter.

Reply:
I'd keep it in the with, it has more context being with the alterresv, and he doesn't need the attrs later.